csrc/core/scalar_type.hpp

```diff
@@ -79,19 +79,74 @@ class ScalarType {
   NanRepr const nan_repr;         // how NaNs are represented
                                   // (not applicable for integer types)
 
-  int64_t size_bits() const { return mantissa + exponent + is_signed(); }
-  bool is_signed() const { return signed_; }
-  bool is_integer() const { return exponent == 0; }
-  bool is_floating_point() const { return exponent > 0; }
-  bool is_ieee_754() const {
+  // unique id for this scalar type that can be computed at compile time for
+  //  c++17 tempalate specialization this is not needed once we migrate to c++20
+  //  and can pass literal classes as template parameters
+  constexpr int64_t id() const {
+    int64_t id = 0;
+    uint32_t bit_offset = 0;
+
+    auto or_and_advance = [&](auto val) {
+      auto constexpr val_size_bits =
+          std::is_same_v<decltype(val), bool> ? 1 : sizeof(val) * 8;
+      id |= (int64_t(val) & ((uint64_t(1) << val_size_bits) - 1)) << bit_offset;
+      bit_offset += val_size_bits;
+    };
+
+    or_and_advance(exponent);
+    or_and_advance(mantissa);
+    or_and_advance(bias);
+    or_and_advance(signed_);
+    or_and_advance(finite_values_only);
+    or_and_advance(nan_repr);
+    return id;
```

Comment:
maybe add a static_assert that `bit_offset <= sizeof(int64_t)*8`?

Reply:
unfortunately `bit_offset` is not a `constexpr` (this function is only `constexpr` if its called on a `constexpr` constructed class, otherwise its runtime function).

This unfortunately makes the check very challenging, but I added a bunch more template magic and I think I got to something reasonable, take a look and let me know what you think :+1:

Reply:
Where is the added template magic?

Reply:
oh I was just referring to `reduce_members` and `reduce_member_types`.